README.md
# n8n-nodes-puppeteer-extended

![n8n.io - Workflow Automation](https://raw.githubusercontent.com/n8n-io/n8n/master/assets/n8n-logo.png)

[n8n](https://www.n8n.io) node based on [n8n-nodes-puppeteer](https://github.com/drudge/n8n-nodes-puppeteer) for requesting webpages using [Puppeteer](https://pptr.dev/), a Node library which provides a high-level API to control Chrome or Chromium over the [DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/).

This version brings features allowing more complex interactions with web pages:

- Click on elements
- Fill in fields
- Fetch content or take a screenshot of a specific area (via CSS selectors)
- Convert HTML to JSON (and turn any page into an API)
- Configure the viewport
- Set additional conditions to validate the loading of a page (ex: wait for selector)
- Inject HTML/CSS/JS
- Allow sequences of actions by combining nodes on a single execution while you keep the same session (navigate on multiple pages, take several screenshots...)
- Allow multiple output combination on a single node
- Granular management of options (globally and on each sequence)

## How to install

### Community Nodes (Recommended)

For users on n8n v0.187+, your instance owner can install this node from [Community Nodes](https://docs.n8n.io/integrations/community-nodes/installation/).

1. Go to **Settings > Community Nodes**.
2. Select **Install**.
3. Enter `n8n-nodes-puppeteer-extended` in **Enter npm package name**.
4. Agree to the [risks](https://docs.n8n.io/integrations/community-nodes/risks/) of using community nodes: select **I understand the risks of installing unverified code from a public source**.
5. Select **Install**.

After installing the node, you can use it like any other node. n8n displays the node in search results in the **Nodes** panel.

### Manual installation

To get started install the package in your n8n root directory:

`npm install n8n-nodes-puppeteer-extended`


For Docker-based deployments, you'll need to make sure [puppeteer is installed](https://developer.chrome.com/docs/puppeteer/troubleshooting/#running-puppeteer-in-docker) first. Then, add the following line before the font installation command in your [n8n Dockerfile](https://github.com/n8n-io/n8n/blob/master/docker/images/n8n/Dockerfile):


`RUN cd /usr/local/lib/node_modules/n8n && npm install n8n-nodes-puppeteer-extended`

Check out [this gist](https://gist.github.com/hckdotng/4232ceaaab6a3533aaca022cc2fdd726) for a working example with docker-compose.

>
> **Note:** If you've having issues running puppeteer, please check their [Troubleshooting guide](https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md) before opening an issue here.
>

## Node Reference

* **Global options**
    * **Emulate Device**: Allows you to specify a [device](https://github.com/puppeteer/puppeteer/blob/main/src/common/DeviceDescriptors.ts) to emulate when requesting the page.
    * **Extra Headers**: Allows you add additional headers when requesting the page.
    * **Launch Arguments**: Allows you to specify additional command line arguments passed to the browser instance.
    * **Viewport**: Allows you to specify a window size in pixels (Width and Height)
    * **Timeout**: Allows you to specify tge maximum navigation time in milliseconds. You can pass 0 to disable the timeout entirely.
    * **Wait Until**: Allows you to change how Puppeteer considers navigation completed.
        * `load`: The load event is fired.
        * `networkidle0`: No more than 0 connections for at least 500 ms.
        * `networkidle2`: No more than 2 connections for at least 500 ms.
    * **Time to Wait**: Allows you specify a time (ms) for Puppeteer to wait before any interactions and content output.
    * **Wait for Selector**: Allows you specify Puppeteer to wait for a selector to appear in page before any interactions and content output.
    * **Page Caching**: Allows you to toggle whether pages should be cached when requesting.
    * **Headless mode**: Allows you to change whether to run browser runs in headless mode or not.
    * **Stealth mode**: When enabled, applies various techniques to make detection of headless Puppeteer harder. Powered by [puppeteer-extra-plugin-stealth](https://github.com/berstend/puppeteer-extra/tree/master/packages/puppeteer-extra-plugin-stealth).
    * **Proxy Server**: Allows Puppeteer to use a custom proxy configuration. You can specify a custom proxy configuration in three ways:
        By providing a semi-colon-separated mapping of list scheme to url/port pairs.
        For example, you can specify:

            http=foopy:80;ftp=foopy2

        to use HTTP proxy "foopy:80" for http URLs and HTTP proxy "foopy2:80" for ftp URLs.

        By providing a single uri with optional port to use for all URLs.
        For example:

            foopy:8080

        will use the proxy at foopy:8080 for all traffic.

        By using the special "direct://" value.

            direct://" will cause all connections to not use a proxy.
    * **Inject HTML**: Allows you to append HTML to the body.
    * **Inject CSS**: Allows you to append raw CSS to the head.
    * **Inject JS**: Allows you to append raw JS to the head.
* **Node options**: These options override any corresponding global options for the step
    * **Timeout**: Allows you to specify tge maximum navigation time in milliseconds. You can pass 0 to disable the timeout entirely.
    * **Wait Until**: Allows you to change how Puppeteer considers navigation completed.
        * `load`: The load event is fired.
        * `networkidle0`: No more than 0 connections for at least 500 ms.
        * `networkidle2`: No more than 2 connections for at least 500 ms.
    * **Time to Wait**: Allows you specify a time (ms) for Puppeteer to wait before any interactions and content output.
    * **Wait for Selector**: Allows you specify Puppeteer to wait for a selector to appear in page before any interactions and content output.
    * **Inject HTML**: Allows you to append HTML to the body.
    * **Inject CSS**: Allows you to append raw CSS to the head.
    * **Inject JS**: Allows you to append raw JS to the head.
* **URL**: URL to the requesting page. Required on the first node but if it's empty in a following node, Puppeteer will stay on the same page. It can be useful to achieve complex operations.
* **Query Parameters**: The query parameter to send.
* **Interactions**: Allows you to click on elements or fill fields, in order Puppeteer will wait for the element to display (timeout 10s) so you can easily chain multiple interactions even if there is asynchonous call.
    * **Selector**: CSS selector to specify the element to interact with.
    * **Value**: If specified, Puppeteer will try to type the value with the keyboard emulation, if not, Puppeteer will click on the element.
    * **Wait for navigation**: Set it to true only if the click triggers a page change, so Puppeteer will wait before any other action.
* **Output**: Output is optional, but if you decide to output data in the current step, you can't select multiple data output even if the UI of n8n allows it. If you need multiple output just add another step without URL to stay on the same page.
    * **Page content**: Returns raw HTML (string) or JSON depending on the options selected and the context.
      * **Property name**: Name of the property returned for this output.
      * **CSS selector**: If empty it will returns the whole page (document), if not the specified element.
      * **Select All**: Instead of grabbing the first element matching the CSS selector, all matching elements will be returned.
      * **innerHTML**: In case of CSS selector (and HTML to JSON set to false), it will returns the raw innerHTML instead of the outerHTML.
      * **HTML to JSON**: Try to convert HTML nodes to JSON with the following logic:
            
            <a href="/contact" class="mr-2 text-sm">Contact</a>

            "a": {
                "@href": "/contact",
                "@class": [
                    "mr-2",
                    "text-sm"
                ],
                "#text": "Contact"
            }

            <div>
                <div class="bg-red-500">
                    <p>Hello</p>
                    <p class="text-xl">Wo<b>r</b>ld!</p>
                </div>
            </div>

            "div": {
                "div": {
                    "@class": "bg-red-500",
                    "p": [
                        "Hello",
                        {
                            "@class": "text-xl",
                            "#text": [
                                "Wo",
                                "ld!"
                            ],
                            "b": "r"
                        }
                    ]
                }
            }

      * **No attributes**: Ignore HTML attributes while converting to JSON.

    * **Screenshot**: Returns binary.
        * **CSS selector**: If empty it captures the whole window or page depending on the options, if not it captures the corresponding zone.
        * **Property Name**: Name of the binary property returned.
        * **Type**: Image type to use between JPG, PNG and WebP.
        * **Quality**: The quality of the image, between 0-100. Not applicable to PNG.
        * **Full Page**: Allows you to capture a screen of the full scrollable content.
    * **PDF**: Returns binary.
        * **Property Name**: Name of the binary property returned.
        * **Page Ranges**: Allows you to specify paper ranges to print, e.g. 1-5, 8, 11-13.
        * **Scale**: Allows you to scale the rendering of the web page. Amount must be between 0.1 and 2
        * **Prefer CSS Page Size**: Give any CSS @page size declared in the page priority over what is declared in the width or height or format option.
        * **Format**: Allows you to specify the paper format types when printing a PDF. eg: Letter, A4.
        * **Height**: Allows you to set the height of paper. You can pass in a number or a string with a unit.
        * **Width**: Allows you to set the width of paper. You can pass in a number or a string with a unit.
        * **Landscape**: Allows you to control whether to show the header and footer
        * **Margin**: Allows you to specify top, left, right, and bottom margin.
        * **Display Header/Footer**: Allows you to specify whether to show the header and footer.
        * **Header Template**: Allows you to specify the HTML template for the print header. Should be valid HTML with the following classes used to inject values into them:
            - `date`: Formatted print date
            - `title`: Document title
            - `url`: Document location
            - `pageNumber` Current page number
            - `totalPages` Total pages in the document
         * **Footer Template**: Allows you to specify the HTML template for the print footer. Should be valid HTML with the following classes used to inject values into them:
            - `date`: Formatted print date
            - `title`: Document title
            - `url`: Document location
            - `pageNumber` Current page number
            - `totalPages` Total pages in the document
        * **Transparent Background**: Allows you to hide the default white background and allows generate PDFs with transparency.
        * **Background Graphic**: Allows you to include background graphics.

## Screenshot

![](images/screenshot.png)

## License

MIT License

Copyright (c) 2022 Nicholas Penree <nick@penree.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog

## Released (2022-10-25 0.1.0)

/!\ Breaking change with the previous version - Futur versions will be more stable

This version is a partial rewrite of the node, it brings a new architecture (inspired by my community node Discord) in which you can keep a Puppeteer session between your nodes during the same workflow execution. So I removed the "steps" paradigm because it is no longer needed.

An API access to n8n is required to check when the workflow execution stops in order to stop Puppeteer too.

### New Features

* Ability to keep a session between nodes
* You can combine as much output as you want on a single node
* You can now specify the property name for the page content output (useful if you output multiple page content on a single node)

### Bug Fixes

### Refactors

* HTML to JSON improvements
* Due to the new architecture, the UI has been simplified

## Released (2022-09-29 0.0.2)

### New Features

* Loop selector "selectAll" in getPageContent
* "No attribute" option: Ignore HTML attributes while converting to JSON

### Bug Fixes

### Refactors

* HTML to JSON imrpovements (apply some filtering to remove useless spaces, line breaks, empty objects...)

nodes/Puppeteer/Puppeteer.node.options.ts
import { INodeProperties, INodeTypeDescription } from "n8n-workflow";

const globalOptions: INodeProperties[] = [
	{
		displayName: "Emulate Device",
		name: "device",
		type: "options",
		default: "",
		typeOptions: {
			loadOptionsMethod: "getDevices",
		},
		required: false,
	},
	{
		displayName: "Executable path",
		name: "executablePath",
		type: "string",
		required: false,
		default: "",
		description: "A path where Puppeteer expects to find the bundled browser.",
	},
	{
		displayName: "Extra Headers",
		name: "headers",
		placeholder: "Add Header",
		type: "fixedCollection",
		typeOptions: {
			multipleValues: true,
		},
		description: "The headers to send.",
		default: {},
		options: [
			{
				name: "parameter",
				displayName: "Header",
				values: [
					{
						displayName: "Name",
						name: "name",
						type: "string",
						default: "",
						description: "Name of the header.",
					},
					{
						displayName: "Value",
						name: "value",
						type: "string",
						default: "",
						description: "Value to set for the header.",
					},
				],
			},
		],
	},
	{
		displayName: "Launch Arguments",
		name: "launchArguments",
		placeholder: "Add Argument",
		type: "fixedCollection",
		typeOptions: {
			multipleValues: true,
		},
		description:
			"Additional command line arguments to pass to the browser instance.",
		default: {},
		options: [
			{
				name: "args",
				displayName: "",
				values: [
					{
						displayName: "Argument",
						name: "arg",
						type: "string",
						default: "",
						description:
							"The command line argument to pass to the browser instance.",
					},
				],
			},
		],
	},
	{
		displayName: "Viewport",
		name: "viewport",
		placeholder: "Add Viewport",
		type: "fixedCollection",
		typeOptions: {
			multipleValues: false,
		},
		default: {},
		options: [
			{
				name: "size",
				displayName: "",
				values: [
					{
						displayName: "Width",
						name: "width",
						type: "number",
						default: 1920,
					},
					{
						displayName: "Height",
						name: "height",
						type: "number",
						default: 1080,
					},
				],
			},
		],
	},
	{
		displayName: "Timeout",
		name: "timeout",
		type: "number",
		typeOptions: {
			minValue: 0,
		},
		default: 30,
		description:
			"Maximum navigation time in milliseconds. Pass 0 to disable timeout.",
	},
	{
		displayName: "Wait Until",
		name: "waitUntil",
		type: "options",
		options: [
			{
				name: "load",
				value: "load",
				description: "The load event is fired",
			},
			{
				name: "networkidle0",
				value: "networkidle0",
				description: "No more than 0 connections for at least 500 ms",
			},
			{
				name: "networkidle2",
				value: "networkidle2",
				description: "No more than 2 connections for at least 500 ms",
			},
		],
		default: "load",
		description: "When to consider navigation succeeded.",
	},
	{
		displayName: "Time to Wait",
		name: "timeToWait",
		type: "number",
		typeOptions: {
			minValue: 0,
		},
		default: 0,
		description: "You can specify a time to wait (ms) before any action",
	},
	{
		displayName: "Wait for Selector",
		name: "waitForSelector",
		type: "string",
		default: "",
		description: "You can specify a time to wait (ms) before any action",
	},
	{
		displayName: "Page Caching",
		name: "pageCaching",
		type: "boolean",
		required: false,
		default: true,
		description: "Whether to enable page level caching. Defaults to true.",
	},
	{
		displayName: "Headless mode",
		name: "headless",
		type: "boolean",
		required: false,
		default: true,
		description: "Whether to run browser in headless mode. Defaults to true.",
	},
	{
		displayName: "Stealth mode",
		name: "stealth",
		type: "boolean",
		required: false,
		default: false,
		description:
			"When enabled, applies various techniques to make detection of headless Puppeteer harder.",
	},
	{
		displayName: "Proxy Server",
		name: "proxyServer",
		type: "string",
		required: false,
		default: "",
		description:
			"This tells Puppeteer to use a custom proxy configuration. Examples: localhost:8080, socks5://localhost:1080, etc.",
	},
	{
		displayName: "Inject HTML",
		name: "injectHtml",
		type: "string",
		required: false,
		typeOptions: {
			rows: 4,
		},
		default: "",
		description: "Custom HTML to inject",
	},
	{
		displayName: "Inject CSS",
		name: "injectCss",
		type: "string",
		required: false,
		typeOptions: {
			rows: 4,
		},
		default: "",
		description: "Custom CSS to inject",
	},
	{
		displayName: "Inject JS",
		name: "injectJs",
		type: "string",
		required: false,
		typeOptions: {
			rows: 4,
		},
		default: "",
		description: "Custom JS to inject",
	},
];

const nodeOptions: INodeProperties[] = [
	{
		displayName: "Timeout",
		name: "timeout",
		type: "number",
		typeOptions: {
			minValue: 0,
		},
		default: 30,
		description:
			"Maximum navigation time in milliseconds. Pass 0 to disable timeout.",
	},
	{
		displayName: "Wait Until",
		name: "waitUntil",
		type: "options",
		options: [
			{
				name: "load",
				value: "load",
				description: "The load event is fired",
			},
			{
				name: "networkidle0",
				value: "networkidle0",
				description: "No more than 0 connections for at least 500 ms",
			},
			{
				name: "networkidle2",
				value: "networkidle2",
				description: "No more than 2 connections for at least 500 ms",
			},
		],
		default: "load",
		description: "When to consider navigation succeeded.",
	},
	{
		displayName: "Time to Wait",
		name: "timeToWait",
		type: "number",
		typeOptions: {
			minValue: 0,
		},
		default: 0,
		description: "You can specify a time to wait (ms) before any action",
	},
	{
		displayName: "Wait for Selector",
		name: "waitForSelector",
		type: "string",
		default: "",
		description: "You can specify a time to wait (ms) before any action",
	},
	{
		displayName: "Inject HTML",
		name: "injectHtml",
		type: "string",
		required: false,
		typeOptions: {
			rows: 4,
		},
		default: "",
		description: "Custom HTML to inject",
	},
	{
		displayName: "Inject CSS",
		name: "injectCss",
		type: "string",
		required: false,
		typeOptions: {
			rows: 4,
		},
		default: "",
		description: "Custom CSS to inject",
	},
	{
		displayName: "Inject JS",
		name: "injectJs",
		type: "string",
		required: false,
		typeOptions: {
			rows: 4,
		},
		default: "",
		description: "Custom JS to inject",
	},
];

const queryParameters: INodeProperties = {
	displayName: "Query Parameters",
	name: "queryParameters",
	placeholder: "Add Parameter",
	type: "fixedCollection",
	typeOptions: {
		multipleValues: true,
	},
	description: "The query parameter to send.",
	default: {},
	options: [
		{
			name: "parameter",
			displayName: "Parameter",
			values: [
				{
					displayName: "Name",
					name: "name",
					type: "string",
					default: "",
					description: "Name of the parameter.",
				},
				{
					displayName: "Value",
					name: "value",
					type: "string",
					default: "",
					description: "Value of the parameter.",
				},
			],
		},
	],
};

const interactions: INodeProperties = {
	displayName: "Interactions",
	name: "interactions",
	placeholder: "Add Interaction",
	type: "fixedCollection",
	typeOptions: {
		multipleValues: true,
	},
	description: "The interaction (click or field filling) to execute",
	default: {},
	options: [
		{
			name: "parameter",
			displayName: "Parameter",
			values: [
				{
					displayName: "Selector",
					name: "selector",
					type: "string",
					default: "",
					description: "Name of the parameter.",
				},
				{
					displayName: "Value (optional)",
					name: "value",
					type: "string",
					default: "",
					description:
						"If empty, Puppeteer will click on the selector, otherwise the value will be filled in the corresponding field",
				},
				{
					displayName: "Wait for navigation",
					name: "waitForNavigation",
					type: "boolean",
					required: false,
					default: false,
					description:
						"If the click will trigger a page loading, set it to true to prevent actions during the loading",
				},
			],
		},
	],
};

const output: INodeProperties = {
	displayName: "Output",
	name: "output",
	placeholder: "Select",
	type: "fixedCollection",
	typeOptions: {
		multipleValues: true,
	},
	description: "You can specify an output on the current node.",
	default: {},
	options: [
		{
			name: "getPageContent",
			displayName: "Page content",
			values: [
				{
					displayName: "Property Name",
					name: "dataPropertyName",
					type: "string",
					required: true,
					default: "pageContent",
					description: "Name of the json key to access the page content.",
				},
				{
					displayName: "CSS selector",
					description:
						"If a CSS selector is specified, the corresponding zone will be output",
					name: "cssSelector",
					type: "string",
					default: "",
				},
				{
					displayName: "Select All",
					name: "selectAll",
					type: "boolean",
					required: false,
					default: false,
					description:
						"Instead of grabbing the first element matching the CSS selector, all matching elements will be returned",
				},
				{
					displayName: "innerHTML",
					name: "innerHtml",
					type: "boolean",
					displayOptions: {
						show: {
							htmlToJson: [false],
						},
					},
					required: false,
					default: false,
					description:
						"By default outerHTML is used for CSS selector. Doesn't apply when HTML to JSON is active",
				},
				{
					displayName: "HTML to JSON",
					name: "htmlToJson",
					type: "boolean",
					required: false,
					default: false,
					description: "Convert HTML nodes to JSON",
				},
				{
					displayName: "No attributes",
					name: "noAttributes",
					type: "boolean",
					displayOptions: {
						show: {
							htmlToJson: [true],
						},
					},
					required: false,
					default: false,
					description: "Ignore HTML attributes while converting to JSON",
				},
			],
		},
		{
			name: "getScreenshot",
			displayName: "Screenshot",
			values: [
				{
					displayName: "Property Name",
					name: "dataPropertyName",
					type: "string",
					required: true,
					default: "data",
					description:
						"Name of the binary property in which  to store the image or PDF data.",
				},
				{
					displayName: "CSS selector",
					description:
						"If a CSS selector is specified, the corresponding zone will be output.",
					name: "cssSelector",
					type: "string",
					default: "",
				},
				{
					displayName: "Type",
					name: "imageType",
					type: "options",
					options: [
						{
							name: "JPEG",
							value: "jpeg",
						},
						{
							name: "PNG",
							value: "png",
						},
						{
							name: "WebP",
							value: "webp",
						},
					],
					default: "png",
					description:
						"The image type to use. PNG, JPEG, and WebP are supported.",
				},
				{
					displayName: "Quality",
					name: "quality",
					type: "number",
					typeOptions: {
						minValue: 0,
						maxValue: 100,
					},
					displayOptions: {
						hide: {
							imageType: ["png"],
						},
					},
					default: 100,
					description:
						"The quality of the image, between 0-100. Not applicable to png images.",
				},
				{
					displayName: "Full Page",
					name: "fullPage",
					type: "boolean",
					required: true,
					default: false,
					description:
						"When true, takes a screenshot of the full scrollable page. If a CSS selector is specified, this parameter will be ignored.",
				},
			],
		},
		{
			name: "getPDF",
			displayName: "PDF",
			values: [
				{
					displayName: "Property Name",
					name: "dataPropertyName",
					type: "string",
					required: true,
					default: "page",
					description:
						"Name of the binary property in which  to store the image or PDF data.",
				},
				{
					displayName: "Page Ranges",
					name: "pageRanges",
					type: "string",
					required: false,
					default: "",
					description: "Paper ranges to print, e.g. 1-5, 8, 11-13.",
				},
				{
					displayName: "Scale",
					name: "scale",
					type: "number",
					typeOptions: {
						minValue: 0.1,
						maxValue: 2,
					},
					default: 1.0,
					required: true,
					description:
						"Scales the rendering of the web page. Amount must be between 0.1 and 2.",
				},
				{
					displayName: "Prefer CSS Page Size",
					name: "preferCSSPageSize",
					type: "boolean",
					required: true,
					default: true,
					description:
						"Give any CSS @page size declared in the page priority over what is declared in the width or height or format option.",
				},
				{
					displayName: "Format",
					name: "format",
					type: "options",
					options: [
						{
							name: "Letter",
							value: "Letter",
							description: "8.5in x 11in",
						},
						{
							name: "Legal",
							value: "Legal",
							description: "8.5in x 14in",
						},
						{
							name: "Tabloid",
							value: "Tabloid",
							description: "11in x 17in",
						},
						{
							name: "Ledger",
							value: "Ledger",
							description: "17in x 11in",
						},
						{
							name: "A0",

							value: "A0",
							description: "33.1in x 46.8in",
						},
						{
							name: "A1",
							value: "A1",
							description: "23.4in x 33.1in",
						},
						{
							name: "A2",
							value: "A2",
							description: "16.54in x 23.4in",
						},
						{
							name: "A3",
							value: "A3",
							description: "11.7in x 16.54in",
						},
						{
							name: "A4",
							value: "A4",
							description: "8.27in x 11.7in",
						},
						{
							name: "A5",
							value: "A5",
							description: "5.83in x 8.27in",
						},
						{
							name: "A6",
							value: "A6",
							description: "4.13in x 5.83in",
						},
					],
					default: "Letter",
					description:
						"Valid paper format types when printing a PDF. eg: Letter, A4",
				},
				{
					displayName: "Height",
					name: "height",
					type: "string",
					default: "",
					required: false,
					description:
						"Sets the height of paper. You can pass in a number or a string with a unit.",
				},
				{
					displayName: "Width",
					name: "width",
					type: "string",
					default: "",
					required: false,
					description:
						"Sets the width of paper. You can pass in a number or a string with a unit.",
				},
				{
					displayName: "Landscape",
					name: "landscape",
					type: "boolean",
					required: true,
					default: true,
					description: "Whether to show the header and footer.",
				},
				{
					displayName: "Margin",
					name: "margin",
					type: "collection",
					placeholder: "Add Margin",
					default: {},
					description: "Set the PDF margins.",
					options: [
						{
							displayName: "Top",
							name: "top",
							type: "string",
							default: "",
							required: false,
						},
						{
							displayName: "Bottom",
							name: "bottom",
							type: "string",
							default: "",
							required: false,
						},
						{
							displayName: "Left",
							name: "left",
							type: "string",
							default: "",
							required: false,
						},
						{
							displayName: "Right",
							name: "right",
							type: "string",
							default: "",
							required: false,
						},
					],
				},
				{
					displayName: "Display Header/Footer",
					name: "displayHeaderFooter",
					type: "boolean",
					required: true,
					default: false,
					description: "Whether to show the header and footer.",
				},
				{
					displayName: "Header Template",
					name: "headerTemplate",
					typeOptions: {
						rows: 5,
					},
					type: "string",
					default: ``,
					description: `HTML template for the print header. Should be valid HTML with the following classes used to inject values into them: - date formatted print date
		
								- title document title
		
								- url document location
		
								- pageNumber current page number
		
								- totalPages total pages in the document`,
					noDataExpression: true,
				},
				{
					displayName: "Footer Template",
					name: "footerTemplate",
					typeOptions: {
						rows: 5,
					},
					type: "string",
					default: ``,
					description: `HTML template for the print footer. Should be valid HTML with the following classes used to inject values into them: - date formatted print date`,
					noDataExpression: true,
				},
				{
					displayName: "Transparent Background",
					name: "omitBackground",
					type: "boolean",
					required: true,
					default: false,
					description:
						"Hides default white background and allows generating pdfs with transparency.",
				},
				{
					displayName: "Background Graphics",
					name: "printBackground",
					type: "boolean",
					required: true,
					default: false,
					description: "Set to true to include background graphics.",
				},
			],
		},
	],
};

/**
 * Options to be displayed
 */
export const nodeDescription: INodeTypeDescription = {
	displayName: "Puppeteer",
	name: "puppeteer",
	group: ["puppeteer"],
	version: 1,
	description: "Request a webpage using Puppeteer",
	defaults: {
		name: "Puppeteer",
		color: "#125580",
	},
	icon: "file:puppeteer.svg",
	inputs: ["main"],
	outputs: ["main"],
	credentials: [
		{
			name: "n8nApi",
			required: true,
		},
	],
	properties: [
		{
			displayName: "Global options",
			name: "globalOptions",
			type: "collection",
			placeholder: "Add Option",
			default: {},
			options: [...globalOptions],
			description:
				"These options must be set on the first puppeteer node in your workflow, they apply to all puppeteer nodes. Any global option set after puppeteer has been initialized will be ignored.",
		},
		{
			displayName: "Node options",
			name: "nodeOptions",
			type: "collection",
			placeholder: "Add Option",
			default: {},
			options: [...nodeOptions],
			description:
				"These options apply only on this node and will override any existing global option.",
		},
		{
			displayName: "URL",
			name: "url",
			type: "string",
			default: "",
			description:
				"Leave the field empty if you want to stay on the current URL (must be set on the first Puppeteer node).",
		},
		{ ...queryParameters },
		{ ...interactions },
		{ ...output },
	],
};
